### ISSUE TITLE
Duplicate entries @ "Recent activities" for (deleted) files

### ISSUE THREAD

**Issue by rickerde** [humhub/humhub#6433]
These are steps that lead to an error:
1. Upload a random PDF File (e.g. abc.pdf) to a random space
2. Then delete this file
3. Upload the same file again.
4. Now log into HumHub with a different user. Under "Recent activities", you will see duplicate entries like "Peter has created the file abc.pdf" twice, still from the first round (although it was deleted) and the newst upload (for the file which is still there).

If a file is deleted, the associated activitiy in the activity stream should vanish also.

| HumHub version  1.14.2 |

**Comment by yurabakhtin** [comment]
@rickerde Sorry, but I cannot reproduce this bug on HumHub version 1.14.2 (May 22, 2023) and module Files version 0.16.0 (April 27, 2023).

I remember similar issues, they probably were fixed here https://github.com/humhub/humhub/pull/6359 for HumHub version 1.14.3 (Not relaesed yet).

If you can give more details with screenshots it would be good, thanks.

**Comment by rickerde** [comment]
Thank you for looking into this. The versions I have installed are slightly different:

- HumHub version **1.14.2**
- Module Files version **0.16.1**

I just reproduced the bug again. Please find as follows the screenshots.
1 I created a file called "abd.pdf",
2 uploaded it
3 deleted it in Humhub
4 uploaded it again
5 then switched to a different user
6 and see two upload entries under recent activities

![step1](https://github.com/humhub/humhub/assets/117586071/5e5cc19c-4ac4-4efa-9153-4caac7ce4978)
![step3](https://github.com/humhub/humhub/assets/117586071/a14b4246-7beb-4196-9b32-bfb0f8ef0ccd)
![step4](https://github.com/humhub/humhub/assets/117586071/ec87da34-709d-4fbf-87b4-c17c90b308b9)

**Comment by rickerde** [comment]
Thank you for looking into this. The versions I have installed are slightly different:

HumHub version 1.14.2
Module Files version 0.16.1

I just reproduced the bug again. Please find as follows the screenshots.
1 I created a file called "abd.pdf",
2 uploaded it
3 deleted it in Humhub
4 uploaded it again
5 then switched to a different user
6 and see two upload entries under recent activities

![step4](https://github.com/humhub/humhub/assets/117586071/8d536d3b-ea51-4495-8757-265128817ce9)
![step1](https://github.com/humhub/humhub/assets/117586071/cd1edb9d-1d6b-4c87-b12b-630ee027a858)
![step3](https://github.com/humhub/humhub/assets/117586071/397a394e-639d-4d66-9278-89c24843bc2a)

**Comment by rickerde** [comment]
Thank you for looking into this. The versions I have installed are slightly different:

HumHub version 1.14.2
Module Files version 0.16.1

I just reproduced the bug again. Please find as follows the screenshots.
1 I created a file called "abc.pdf",
2 uploaded it
3 deleted it in Humhub
4 uploaded it again
5 then switched to a different user
6 and then see two upload entries under recent activities

![step1](https://github.com/humhub/humhub/assets/117586071/cd1edb9d-1d6b-4c87-b12b-630ee027a858)
![step3](https://github.com/humhub/humhub/assets/117586071/397a394e-639d-4d66-9278-89c24843bc2a)
![step4](https://github.com/humhub/humhub/assets/117586071/8d536d3b-ea51-4495-8757-265128817ce9)
![step4](https://github.com/humhub/humhub/assets/117586071/d6244455-7f80-47f0-b67d-e9332111fcb2)
![step1](https://github.com/humhub/humhub/assets/117586071/8543da9a-1826-40fa-bbf7-1cca0da954a7)
![step3](https://github.com/humhub/humhub/assets/117586071/92625ece-c510-445a-855d-9139987f4ca4)

**Comment by rickerde** [comment]
Thank you for looking into this. The versions I have installed are slightly different:

HumHub version 1.14.2
Module Files version 0.16.1

I just reproduced the bug again. Please find as follows the screenshots.
1 I created a file called "abc.pdf",
2 uploaded it
3 deleted it in Humhub
4 uploaded it again
5 then switched to a different user
6 and then see two upload entries under recent activities

![step1](https://github.com/humhub/humhub/assets/117586071/c32b6cfd-7590-4c60-ab11-9557b3c1aebe)
![step3](https://github.com/humhub/humhub/assets/117586071/ba140888-8587-4ab4-8dd7-15dfcd1cb95c)
![step4](https://github.com/humhub/humhub/assets/117586071/6b28846b-cd86-41dc-9432-ba7cd89b4eba)

**Comment by yurabakhtin** [comment]
@rickerde Thank you for additional details, it helped me to find the issue, will be fixed in next HumHub release `1.14.3`.

@luke- I have decided to delete all activites of `ContentActiveRecord` when the record is updated to not published state, because on each next publishing we again run a code to create new activities, so it would be better to keep only activity right after publishing instead of restore old activity woth old date, i.e. now after publishing/restoring a content user will see more correct activity like "User Name created a new File 1 minute ago" instead of previous wrong "User Name created a new File 3 days ago".
PR https://github.com/humhub/humhub/pull/6443.

**Comment by luke-** [comment]
@yurabakhtin If I understand it correctly, the problem affects mainly the File module, because there, when creating a file with the same name and not yet `hardDeleted` - the old record is used. Then I would only delete there the old activities beforehand in such cases.  

Instead of always in core deleting all activities directly. Then this could also be restored with.

Or am I missing something here?

**Comment by yurabakhtin** [comment]
@luke- The problem is related not only for the "Files" module. This bug can be related for all code which restore a Content from not Published state, i.e. if a Content was published before and then it was modified to "Deleted" or other state like "Draft" then on next publishing we have the issue because new Activity is created.
We can reproduce this even with simple Post record by module https://github.com/humhub/recycle-bin:

- Create a Post as Published
- Delete the Post
- Go to "Deleted" filter
- Use action "Restore" of the module "Recycle Bin"
- Result: two activities: first - "PostX created 10 minutes ago", second - "PostX created 1 minute ago"
- Expexted: single activity: "PostX created 1 minute ago"

> Instead of always in core deleting all activities directly. Then this could also be restored with.

What reason to keep Activities of the soft deleted content?
If we will keep them in DB then it would be better to update their date to current time instead of use the old date of real creating time. If yes then I think it is easier to remove all Activities on soft deletion and then create new ones when a Content will be published.

Also currently we create an Activity record only when Conent record is created as Published, i.e. if we will create Draft or Scheduled post we don't create Activity, we create it only on Publishing. So it make sense to delete Activity when we modify a Conent record to Draft, Scheduled or Deleted state.

**Comment by luke-** [comment]
> @luke- The problem is related not only for the "Files" module. This bug can be related for all code which restore a Content from not Published state, i.e. if a Content was published before and then it was modified to "Deleted" or other state like "Draft" then on next publishing we have the issue because new Activity is created. We can reproduce this even with simple Post record by module https://github.com/humhub/recycle-bin:
> 
> * Create a Post as Published
> * Delete the Post
> * Go to "Deleted" filter
> * Use action "Restore" of the module "Recycle Bin"
> * Result: two activities: first - "PostX created 10 minutes ago", second - "PostX created 1 minute ago"
> * Expexted: single activity: "PostX created 1 minute ago"

I would expect one single activity "PostX created 10 minutes ago" and no new "Created" activity. 

> > Instead of always in core deleting all activities directly. Then this could also be restored with.
> 
> What reason to keep Activities of the soft deleted content? If we will keep them in DB then it would be better to update their date to current time instead of use the old date of real creating time. If yes then I think it is easier to remove all Activities on soft deletion and then create new ones when a Content will be published.

In case of a restore using the "Recycle Bin" module, it would be rather nice if related activities of a content could be restored as well. 

> Also currently we create an Activity record only when Conent record is created as Published, i.e. if we will create Draft or Scheduled post we don't create Activity, we create it only on Publishing. So it make sense to delete Activity when we modify a Conent record to Draft, Scheduled or Deleted state.

Hmm, yes there are indeed different cases where different handling of the activities makes sense. 
Maybe it makes sense to create an overview of the different cases in the wiki?

**Comment by yurabakhtin** [comment]
@luke- 

> I would expect one single activity "PostX created 10 minutes ago" and no new "Created" activity.

I think in case of auto-restoring a File by the module "Files" it will be unexpected for many users, for example:

- First user uploaded a File with name "abc.pdf" 2 years ago,
- Then the file was deleted(softly) 1 year ago by someone,
- Then second user uploaded another File with the same name "abc.pdf" today,
- I think the second user expects to see an activity "File created today" instead of "File created 2 years ago", and it seems the old activity will not be visible because of sorted in the end of Activities list.

> In case of a restore using the "Recycle Bin" module, it would be rather nice if related activities of a content could be restored as well.

Maybe yes, but I think here as for "Files" modules the old Activities will be hidden in the end of Activities list, or we should implement new Activity with text like "PostX restored 1 minute ago".

> Hmm, yes there are indeed different cases where different handling of the activities makes sense.
> Maybe it makes sense to create an overview of the different cases in the wiki?

Ok, will you create the wiki page for this?

**Comment by yurabakhtin** [comment]
@luke- 

> In the case of the File module you are right.
> 
> Probably we should change the handling here. So if a file is in status Pending/Soft Deletion and in the meantime a new file with the same filename is uploaded, we could also rename the deleted file e.g. from "test.doc" to "conflict[fileid]-test.doc". (Same for other modules)
> 
> This would also solves some other problems (e.g. OnlyOffice module [ONLYOFFICE/onlyoffice-humhub#100](https://github.com/ONLYOFFICE/onlyoffice-humhub/pull/100))

Ok, as I understood we should modify **core** code, (not CFiles module):

- If we new `File` is uploaded with same name and its `ContentActiveRecord` is soft deleted then we should rename to `conflict[fileid]-test.doc`, (fileid - is ID of the soft deleted File, right?)
- otherwise do as now.

Should I create a separate issue for this or can I start to implement it here?

---

> Such an activity would be interesting. Maybe something for the Recycle Bin module?

Yes, I can implement new Activity from the "Recycle Bin" module side.
But I am thinking about this code https://github.com/humhub/humhub/blob/master/protected/humhub/modules/content/models/Content.php#L241-L243:

```php
if ($this->getStateService()->isPublished()) {
    // Run process for new content(Send notifications) only after publishing the Content
    $this->processNewContent();
```

I think firstly this code should run only when previous State was not `Deleted` otherwise the Activity(New created content) will be created on each restoring. It is not easy to be handled from module side.

---

> https://community.humhub.com/s/contribution-core-development/wiki/Stateable

Yes, I agree with that, so this point should be fixed in core as I wrote above, i.e. the condition `if ($this->getStateService()->isPublished()) {` should be changed to `if ($previousState !== Content::STATE_DELETED && $this->getStateService()->isPublished()) {`:

- DELETED => PUBLISHED | Do not fire Mentioning or default created Notifications/Activities

Should I fix this in core?
If yes, then as I understood we don't neet the changes from here https://github.com/humhub/humhub/pull/6443/files, and I can use this PR to revert them and apply new with `previousState !== Content::STATE_DELETED`.

**Comment by luke-** [comment]
> @luke-
> 
> > In the case of the File module you are right.
> > Probably we should change the handling here. So if a file is in status Pending/Soft Deletion and in the meantime a new file with the same filename is uploaded, we could also rename the deleted file e.g. from "test.doc" to "conflict[fileid]-test.doc". (Same for other modules)
> > This would also solves some other problems (e.g. OnlyOffice module [ONLYOFFICE/onlyoffice-humhub#100](https://github.com/ONLYOFFICE/onlyoffice-humhub/pull/100))
> 
> Ok, as I understood we should modify **core** code, (not CFiles module):
> 
> * If we new `File` is uploaded with same name and its `ContentActiveRecord` is soft deleted then we should rename to `conflict[fileid]-test.doc`, (fileid - is ID of the soft deleted File, right?)
> * otherwise do as now.
> 
> Should I create a separate issue for this or can I start to implement it here?

No. as I understood, the CFiles module has implemented a special handling here.
Specifically, when a file is marked as `Deleted` it get the status `Published` again if they have the same filename (`CFiles.Title`). 

e.g. https://github.com/humhub/cfiles/blob/master/controllers/EditController.php#L52

Instead of restoring the file/folder here. We should rename the soft deleted file/folder and create a fresh record as normal. 

> 
> > Such an activity would be interesting. Maybe something for the Recycle Bin module?
> 
> Yes, I can implement new Activity from the "Recycle Bin" module side. But I am thinking about this code https://github.com/humhub/humhub/blob/master/protected/humhub/modules/content/models/Content.php#L241-L243:
> 
> ```
> if ($this->getStateService()->isPublished()) {
>     // Run process for new content(Send notifications) only after publishing the Content
>     $this->processNewContent();
> ```
> 
> I think firstly this code should run only when previous State was not `Deleted` otherwise the Activity(New created content) will be created on each restoring. It is not easy to be handled from module side.
> 

Ok, yes it's important that we don't run `processNewContent()` more than one time. Maybe we should rename this method or improve the PHPDoc to reflect that.

Any reason to always execute this method expect the old status was `DELETED`?
It should not be possible to switch from `PUBLISHED` back to `DRAFT` or `SCHEDULED`.  

---  

Regarding the `Restore Activity` this should be quite easily doable on module side? I would suggest to create an issue for this in the Recycle bin module. We can implement this later.

> > https://community.humhub.com/s/contribution-core-development/wiki/Stateable
> 
> Yes, I agree with that, so this point should be fixed in core as I wrote above, i.e. the condition `if ($this->getStateService()->isPublished()) {` should be changed to `if ($previousState !== Content::STATE_DELETED && $this->getStateService()->isPublished()) {`:
> 
> * DELETED => PUBLISHED | Do not fire Mentioning or default created Notifications/Activities
> 
> Should I fix this in core? If yes, then as I understood we don't neet the changes from here https://github.com/humhub/humhub/pull/6443/files, and I can use this PR to revert them and apply new with `previousState !== Content::STATE_DELETED`.

Yes, I think we should revert 6443. What do you think?

**Comment by yurabakhtin** [comment]
@luke- 

> No. as I understood, the CFiles module has implemented a special handling here. Specifically, when a file is marked as `Deleted` it get the status `Published` again if they have the same filename (`CFiles.Title`).
> 
> e.g. https://github.com/humhub/cfiles/blob/master/controllers/EditController.php#L52
> 
> Instead of restoring the file/folder here. We should rename the soft deleted file/folder and create a fresh record as normal.

Ok, can I start to modify the Cfiles module for the cases when file/folder is restoring from `Deleted` state.
Also I see the CFiles module uses deprecated `Content::setState` and other old code like `$file->content->state === Content::STATE_DELETED`, so firstly we should update the module use new `Content->getStateService()->...`.

---

> Ok, yes it's important that we don't run `processNewContent()` more than one time. Maybe we should rename this method or improve the PHPDoc to reflect that.
> 
> Any reason to always execute this method expect the old status was `DELETED`? It should not be possible to switch from `PUBLISHED` back to `DRAFT` or `SCHEDULED`.

We implemented to execute the method `processNewContent()` for cases when Content was created as `DRAFT` or `SCHEDULED` and then on publishing we need to run it, so the Content looks like created recently at the publishing time.
Yes, maybe we should not run the method when status was `DELETED`, but please imagine a case:

- Post is created as `DRAFT` or `SCHEDULED`,
- Post was deleted without publishing, so the method was not executed,
- Post was restored to `PUBLISHED` state, so no any expected activites created.

Theoretically some external modules may change any state to any state, so I am thinking how to avoid double activities from core side and don't skip when a Content is published first time.
If we look at the activities table:

![activity_table](https://github.com/humhub/humhub/assets/6995524/8acd2dbf-ff25-459f-a8a6-f0bd550d9314)

What is reason to keep two same records in the table?
What if before creating of any Activity record we will find the a record with same `class`+`module`+`object_model`+`object_id` and use the existing Activity instead of creating new one? If you agree this then I think it would be better to update the column `content.created_at` of the related Activity record.

- Another solution is to keep some new column for a Content record like `was_published = 1`, so we will know that the Content was published once in past and no need to run a related code twice.
- Or we could store all previous used states like `used_states = '1,20,100'`, so we can check the state already was used for the Content with new code like `$content->getStateService()->wasPublished()` or `$content->getStateService()->wasDraft()`.

---

> Yes, I think we should revert 6443. What do you think?

Ok, I will close the PR, because it seems we need completely different solution.

**Comment by luke-** [comment]
> @luke-
> 
> > No. as I understood, the CFiles module has implemented a special handling here. Specifically, when a file is marked as `Deleted` it get the status `Published` again if they have the same filename (`CFiles.Title`).
> > e.g. https://github.com/humhub/cfiles/blob/master/controllers/EditController.php#L52
> > Instead of restoring the file/folder here. We should rename the soft deleted file/folder and create a fresh record as normal.
> 
> Ok, can I start to modify the Cfiles module for the cases when file/folder is restoring from `Deleted` state?
> 
> Also I see the CFiles module uses deprecated `Content::setState` and other old code like `$file->content->state === Content::STATE_DELETED`, so firstly we should update the module use new `Content->getStateService()->...`. (**update:** we already have such PR [humhub/cfiles#178](https://github.com/humhub/cfiles/pull/178), it would be better to merge this before start new changes in the related code)

Ok, yes you can start with this change. I've just merged the open PR.

> > Ok, yes it's important that we don't run `processNewContent()` more than one time. Maybe we should rename this method or improve the PHPDoc to reflect that.
> > Any reason to always execute this method expect the old status was `DELETED`? It should not be possible to switch from `PUBLISHED` back to `DRAFT` or `SCHEDULED`.
> 
> We implemented to execute the method `processNewContent()` for cases when Content was created as `DRAFT` or `SCHEDULED` and then on publishing we need to run it, so the Content looks like created recently at the publishing time. Yes, maybe we should not run the method when status was `DELETED`, but please imagine a case:
> 
> * Post is created as `DRAFT` or `SCHEDULED`,
> * Post was deleted without publishing, so the method was not executed,
> * Post was restored to `PUBLISHED` state, so no any expected activites created.
> 
> Theoretically some external modules may change any state to any state, so I am thinking how to avoid double activities from core side and don't skip when a Content is published first time. If we look at the activities table:
> 
> ![activity_table](https://user-images.githubusercontent.com/6995524/255122441-8acd2dbf-ff25-459f-a8a6-f0bd550d9314.png)
> 
> What is reason to keep two same records in the table? What if before creating of any Activity record we will find the a record with same `class`+`module`+`object_model`+`object_id` and use the existing Activity instead of creating new one? If you agree this then I think it would be better to update the column `content.created_at` of the related Activity record.
> 
> * Another solution is to keep some new column for a Content record like `was_published = 1`, so we will know that the Content was published once in past and no need to run a related code twice.
> * Or we could store all previous used states like `used_states = '1,20,100'`, so we can check the state already was used for the Content with new code like `$content->getStateService()->wasPublished()` or `$content->getStateService()->wasDraft()`.

Oh, good point. We should definitly handle such cases.

What do you think if we introduce a  `wasPublishedBefore()` method which checks this by testing if a `ContentCreated` activity already exists. We can later change this implementation with `used_states` or `was_published` if required.

> > Yes, I think we should revert 6443. What do you think?
> 
> Ok, I will close the PR, because it seems we need completely different solution.

**Comment by yurabakhtin** [comment]
> What do you think if we introduce a `wasPublishedBefore()` method which checks this by testing if a `ContentCreated` activity already exists. We can later change this implementation with `used_states` or `was_published` if required.

@luke- Yes, it can be as temp solution. I think the method name `wasPublished()` is enough, without `Before`.

> Ok, yes you can start with this change. I've just merged the open PR.

Ok I will do this in next PR.

**Comment by luke-** [comment]
> > What do you think if we introduce a `wasPublishedBefore()` method which checks this by testing if a `ContentCreated` activity already exists. We can later change this implementation with `used_states` or `was_published` if required.
> 
> @luke- Yes, it can be as temp solution. I think the method name `wasPublished()` is enough, without `Before`.

Yes, thats good. The solution should be quite stable, right? Or missed I something here.

> 
> > Ok, yes you can start with this change. I've just merged the open PR.
> 
> Ok I will do this in next PR.

**Comment by yurabakhtin** [comment]
> > > What do you think if we introduce a `wasPublishedBefore()` method which checks this by testing if a `ContentCreated` activity already exists. We can later change this implementation with `used_states` or `was_published` if required.
> > 
> > 
> > @luke- Yes, it can be as temp solution. I think the method name `wasPublished()` is enough, without `Before`.
> 
> Yes, thats good. The solution should be quite stable, right? Or missed I something here.

@luke- I see the record `humhub\modules\content\activities\ContentCreated` is created after each publishing of the `humhub\modules\content\models\Content` so we can consider an existing of record `ContentCreated` as flag that `Content` was published at least one time. Therefore I think this solution can be stable.

**Comment by yurabakhtin** [comment]
> > > No. as I understood, the CFiles module has implemented a special handling here. Specifically, when a file is marked as `Deleted` it get the status `Published` again if they have the same filename (`CFiles.Title`).
> > > e.g. https://github.com/humhub/cfiles/blob/master/controllers/EditController.php#L52
> > > Instead of restoring the file/folder here. We should rename the soft deleted file/folder and create a fresh record as normal.
> > 
> > 
> > Ok, can I start to modify the Cfiles module for the cases when file/folder is restoring from `Deleted` state?
> > Also I see the CFiles module uses deprecated `Content::setState` and other old code like `$file->content->state === Content::STATE_DELETED`, so firstly we should update the module use new `Content->getStateService()->...`. (**update:** we already have such PR [humhub/cfiles#178](https://github.com/humhub/cfiles/pull/178), it would be better to merge this before start new changes in the related code)
> 
> Ok, yes you can start with this change. I've just merged the open PR.

@luke- PR https://github.com/humhub/cfiles/pull/186 - I see you already merged the PR, please also merge this one https://github.com/humhub/cfiles/pull/187 for fixing tests.

**Comment by yurabakhtin** [comment]
> > > What do you think if we introduce a `wasPublishedBefore()` method which checks this by testing if a `ContentCreated` activity already exists. We can later change this implementation with `used_states` or `was_published` if required.
> > 
> > 
> > @luke- Yes, it can be as temp solution. I think the method name `wasPublished()` is enough, without `Before`.
> 
> Yes, thats good. The solution should be quite stable, right? Or missed I something here.

@luke- Done in PR https://github.com/humhub/humhub/pull/6456.
